openstreetmap-carto.lua

```diff
@@ -303,134 +575,173 @@ function filter_tags_generic(tags)
+            return
+        end
 
-    local type = keyvalues["type"]
+        local area_tags = isarea(object.tags)
+        if object.is_closed and area_tags then
+            add_polygon(object.tags)
 
-    -- Remove type key
-    keyvalues["type"] = nil
+            if z_order(object.tags) ~= nil then
+                add_transport_polygon(object.tags)
+            end
+        else
+            add_line(object.tags)
 
-    -- Filter out relations with just a type tag or no tags
-    if next(keyvalues) == nil then
-        return 1, keyvalues, members_superseded, 0, 0, 0
-    end
+            if z_order(object.tags) ~= nil then
+                add_transport_line(object.tags)
+            end
 
-    if type == "boundary" or (type == "multipolygon" and keyvalues["boundary"]) then
-        keyvalues.z_order = z_order(keyvalues)
-        return 0, keyvalues, members_superseded, 1, 0, roads(keyvalues)
-    -- For multipolygons...
-    elseif (type == "multipolygon") then
-        -- Multipolygons by definition are polygons, so we know roads = linestring = 0, polygon = 1
-        keyvalues.z_order = z_order(keyvalues)
-        return 0, keyvalues, members_superseded, 0, 1, 0
-    elseif type == "route" then
-        keyvalues.z_order = z_order(keyvalues)
-        return 0, keyvalues, members_superseded, 1, 0, roads(keyvalues)
+            if roads(object.tags) then
+                add_roads(object.tags)
+            end
+        end
+    elseif osm2pgsql.stage == 2 then
+        -- Stage two processing is called on ways that are part of admin boundary relations
+        local props = phase2_admin_ways[object.id]
+        if props ~= nil then
+            tables.admin:add_row({admin_level = props.level, multiple_relations = (props.parents > 1), geom = { create = 'line' }})
+        end
     end
-
-    -- Unknown type of relation or no type tag
-    return 1, keyvalues, members_superseded, 0, 0, 0
 end
 
---- Check if an object with given tags should be treated as polygon
--- @param tags OSM tags
--- @return 1 if area, 0 if linear
-function isarea (tags)
-    -- Treat objects tagged as area=yes polygon, other area as no
-    if tags["area"] then
-        return tags["area"] == "yes" and 1 or 0
+function osm2pgsql.process_relation(object)
+    -- grab the type tag before filtering tags
+    local type = object.tags.type
+    object.tags.type = nil
+
+    if clean_tags(object.tags) then
+        return
     end
+    if type == "boundary" or (type == "multipolygon" and object.tags["boundary"]) then
+        add_line(object.tags)
 
-   -- Search through object's tags
-    for k, v in pairs(tags) do
-        -- Check if it has a polygon key and not a linestring override, or a polygon k=v
-        for _, ptag in ipairs(polygon_keys) do
-            if k == ptag and v ~= "no" and not (linestring_values[k] and linestring_values[k][v]) then
-                return 1
-            end
+        if roads(object.tags) then
```

Comment:
@pnorman 

Shouldn't there be a

`            if z_order(object.tags) ~= nil then
                add_transport_line(object.tags)
            end`

section in this part of the code of the 'osm2pgsql.process_relation' function?

I am asking this because I just discovered an issue in my rendering results using this flex style that I think may be related. Notice how in the below images, sections of the primary roads are missing. The missing sections are dual tagged with boundary=administrative:

![afbeelding](https://user-images.githubusercontent.com/7635726/141824056-010887af-ef75-45fd-af35-5e1369334168.png)

Reply:
My first naive attempt at fixing this, based on the suggested change in the previous comment, seems to fail. I still miss the lines in 'planet_osm_transport_line' table:

![afbeelding](https://user-images.githubusercontent.com/7635726/141848315-061a4d4d-676a-41bf-8870-267c632ba174.png)

Reply:
I think I begin to understand why it failed. As your code comments in the Lua file clearly state, the z-order function returns nil for z=0, which is the value set for the boundary=administrative tag in the local variable 'roads_info'.

Obviously, adding

`if z_order(object.tags) ~= nil then add_transport_line(object.tags) end`

hence fails....

Reply:
This is beginning to look like both the **osm2pgsql.process_way** and o**sm2pgsql.process_relation** function in the Lua style file may need adjustment.

Reply:
@pnorman , I did another interesting observation: when I disable the creation of the 'planet_osm_admin' table, and thus prevent the 'stage2' osm2pgsql processing being triggered, the missing ways do appear in the 'planet_osm_transport_line' table. This is without any changes to the code like adding the extra code lines I suggested.

![afbeelding](https://user-images.githubusercontent.com/7635726/142028761-a414ff5c-0e76-4f80-bb9a-07aa6618b1e7.png)